Save / Load Implemented

Gabeld opened the pull request.

Gabeld committed:
Save / Load Implemented

v-rusu commented on WeatherApp/CityManager.swift:
```diff
@@ -20,9 +20,27 @@ class CityManager {
+            try data.write(to: cityArchiveURL)
+            
+        } catch {
+            fatalError("Error encoding \(error)")
```
don't do fatalErrors

v-rusu replied:
check below also

v-rusu commented on WeatherApp/CityManager.swift:
```diff
@@ -20,9 +20,27 @@ class CityManager {
+        print("Saving user data to \(cityArchiveURL)")
+        do {
+           let data = try NSKeyedArchiver.archivedData(withRootObject: cities, requiringSecureCoding: false)
+            try data.write(to: cityArchiveURL)
```
indenting

v-rusu commented on WeatherApp/DetailedViewController.swift:
```diff
@@ -44,16 +48,37 @@ class DetailedViewController: UIViewController, UICollectionViewDataSource, UICo
+        let topConstraint = detailedCollectionView.topAnchor.constraint(equalTo: view.topAnchor)
+        topConstraint.isActive = true
+        let botConstraint = stackView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
+        botConstraint.isActive = true
```
why aren't these constraints set in the storyboard?

v-rusu commented on WeatherApp/DetailedViewController.swift:
```diff
@@ -146,4 +146,34 @@ class DetailedViewController: UIViewController, UICollectionViewDataSource, UICo
+    }
+    
+    func updateForecast() {
+        
```
why the empty line?

v-rusu commented on WeatherApp/ListCollectionViewCell.swift:
```diff
@@ -42,6 +42,7 @@ class ListCollectionViewCell: UICollectionViewCell, WeatherDataRendering {
         
         layer.cornerRadius = 7
         layer.masksToBounds = true
+        cityNameLabel.lineBreakMode = .byCharWrapping
```
set this in storyboard

v-rusu commented on WeatherApp/WeatherHandler.swift:
```diff
@@ -34,7 +34,80 @@ class WeatherHandler {
+        let request = URLRequest(url: cityURL)
+        let task = URLSession.shared.dataTask(with: request) {
+            (data, response, error) in
+            
```
move "(data, response, error) in" inline with "URLSession.shared.dataTask(with: request) {"
remove empty line

v-rusu commented on WeatherApp/WeatherHandler.swift:
```diff
@@ -34,7 +34,80 @@ class WeatherHandler {
+                    }
+                    
+                    let date = jsonDictionary["dt"]
+                    let weatherDate = Date(timeIntervalSince1970: date!.doubleValue)
```
you should be careful with forceunwrapping

v-rusu commented on WeatherApp/WeatherHandler.swift:
```diff
@@ -34,7 +34,80 @@ class WeatherHandler {
+                    let weatherDate = Date(timeIntervalSince1970: date!.doubleValue)
+                    let weatherArray = jsonDictionary["weather"] as? [[String: AnyObject]]
+                    let weatherDictionary = weatherArray?.first
+                    let weatherCondition = weatherDictionary!["main"] as!  String
```
you should be careful with forceunwrapping

v-rusu replied:
also check lines below

v-rusu commented on WeatherApp/WeatherHandler.swift:
```diff
@@ -34,7 +34,80 @@ class WeatherHandler {
+                    let weatherArray = jsonDictionary["weather"] as? [[String: AnyObject]]
+                    let weatherDictionary = weatherArray?.first
+                    let weatherCondition = weatherDictionary!["main"] as!  String
+                    let conditionID = weatherDictionary!["id"] as! Int
```
you should be careful with forceunwrapping

v-rusu commented on WeatherApp/WeatherHandler.swift:
```diff
@@ -34,7 +34,80 @@ class WeatherHandler {
+                    }
+                    
+                }
+                catch let error {
```
catch inline with }

v-rusu commented on WeatherApp/WeatherHandler.swift:
```diff
@@ -34,7 +34,80 @@ class WeatherHandler {
+        task.resume()
+    }
+    
+    func weatherForCity(city: City) {
```
this should be named: updateWeather(forCity city: City)

weatherForCity implies that the function returns the weather.

v-rusu commented on WeatherApp/WeatherHandler.swift:
```diff
@@ -71,25 +138,13 @@ class WeatherHandler {
                         let weatherData = WeatherData(date: weatherDate, avgTemp: currTemperature, minTemp: minTemperature, maxTemp: maxTemperature, condition: weatherCondition, conditionID: conditionID)
                         
-                        weatherDataList.append(weatherData)
+                        city.weatherData.append(weatherData)
```
shouldn't you clear city.weatherData before appending? 
this could easily pile up duplicate weatherData if called multiple times for the same city

v-rusu replied:
It seems it dupes data based on your usage of the weatherForCity function

v-rusu replied:
I am not sure of the changes you made here, but it seems that at some point you had the correct implementation for this.

Gabeld committed:
Fixed duplicate weatherData

v-rusu commented on WeatherApp/DetailedViewController.swift:
```diff
@@ -44,16 +48,31 @@ class DetailedViewController: UIViewController, UICollectionViewDataSource, UICo
+                detailedCollectionView.scrollToItem(at: indexPath, at: .left, animated: true)
+            }
         }
+        updateForecast()
```
This call is probably useless. You already call it in scrollViewDidScroll

v-rusu commented on WeatherApp/DetailedViewController.swift:
```diff
@@ -80,31 +99,14 @@ class DetailedViewController: UIViewController, UICollectionViewDataSource, UICo
-        }
+        pageController.currentPage = Int(offSet + horizontalCenter) / Int(width)
+        getDataForCurrentCity()
+        updateForecast()
```
These 2 functions will get called a lot of times, might even cause the app to run slow on older devices. You should only make these 2 calls when the displayed city changes.

v-rusu commented on WeatherApp/WeatherHandler.swift:
```diff
@@ -34,34 +34,97 @@ class WeatherHandler {
+        task.resume()
+    }
+    
+    func updateWeather(forCity city: City) {
```
There is no way of knowing when the weather was updated. That means that you cannot update the UI when weather gets updated. You should implement a callback

Gabeld committed:
implemented Callback function to update the UI
fixed performance issue - scrollView will update only when the displayed object is changed

v-rusu commented on WeatherApp/WeatherHandler.swift:
```diff
@@ -93,4 +158,33 @@ class WeatherHandler {
         task.resume()
     }
+    
+    func getWeatherForCities(completion: @escaping (_ result: Bool) -> Void) {
```
you should only call completion block once, after all the cities data was updated.

v-rusu commented on WeatherApp/ListViewController.swift:
```diff
@@ -19,17 +19,29 @@ class ListViewController: UIViewController, UICollectionViewDataSource, UICollec
+    
+    @objc func didBecomeActive(notification: NSNotification!) {
+        // do whatever you want when the app is brought back to the foreground
+        weatherHandler.getWeatherForCities { (succes) in
```
fix typo "succes"

v-rusu commented on WeatherApp/ListViewController.swift:
```diff
@@ -19,17 +19,29 @@ class ListViewController: UIViewController, UICollectionViewDataSource, UICollec
-            weatherHandler.weatherForCity(city: city) { (succes) in
-                self.collectionView.reloadData()
-            }
+        weatherHandler.getWeatherForCities { (succes) in
```
fix typo "succes"
